### PR TITLE
Default process user to container config user.

### LINKED ISSUE #14115: exec doesn't default to user specified in image anymore
In Docker <= 1.6.2, using `docker exec` defaults to the user specified in the image via `USER`.

In Docker 1.7.0, it defaults to `root`. Is this intended or a bug?

docker info:
```
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirperm1 Supported: true
Execution Driver: native-0.2
Kernel Version: 3.16.0-37-generic
Operating System: Ubuntu 14.10
```

1.6.2 docker version:
```
Client version: 1.6.2
Client API version: 1.18
Go version (client): go1.4.2
Git commit (client): 7c8fca2
OS/Arch (client): linux/amd64
Server version: 1.6.2
Server API version: 1.18
Go version (server): go1.4.2
Git commit (server): 7c8fca2
OS/Arch (server): linux/amd64
```

1.7.0 docker version:
```
Client version: 1.7.0
Client API version: 1.19
Go version (client): go1.4.2
Git commit (client): 0baf609
OS/Arch (client): linux/amd64
Server version: 1.7.0
Server API version: 1.19
Go version (server): go1.4.2
Git commit (server): 0baf609
OS/Arch (server): linux/amd64
```

Dockerfile:
```
FROM ubuntu:14.04
RUN   useradd -m -s /bin/bash default
USER  default
```

Steps to reproduce:
```
docker exec -it <container> bash
```

### PR DESCRIPTION
**Pull request by calavera**
Fixes #14115

Signed-off-by: David Calavera david.calavera@gmail.com

### CODE REVIEW

File: daemon/exec.go [hunk 1]
```diff
@@ -231,7 +231,12 @@ func (d *Daemon) ContainerExecStart(execName string, stdin io.ReadCloser, stdout
 }
 
 func (d *Daemon) Exec(c *Container, execConfig *execConfig, pipes *execdriver.Pipes, startCallback execdriver.StartCallback) (int, error) {
-	exitStatus, err := d.execDriver.Exec(c.command, &execConfig.ProcessConfig, pipes, startCallback)
+	pc := &execConfig.ProcessConfig
```
Comment (icecrime) [inline daemon/exec.go]: Maybe not a problem, but isn't this modifying the `execConfig`? Isn't a copy desirable?

**Comment by cpuguy83** [comment]
Looks like it should be ok, but really needs a test.
Probably two tests, one to make sure the user in the exec is as expected, and another that subsequent exec's default user config are correct.

**Comment by calavera** [comment]
do we have a small image to test user changes? I'm not sure if we can test this with busybox.

**Comment by cpuguy83** [comment]
How come?

```
docker run -d -u 1000 --name test busybox top
docker exec test id                   # should be `uid=1000 gid=0(root)`
docker exec -u 1001 test id           # should be `uid=1001 gid=0(root)`
docker exec test id                   # should be `uid=1000 gid=0(root)`
```

**Comment by calavera** [comment]
I added a test

File: daemon/exec.go (hunk 1 above)
Comment (LK4D4) [inline daemon/exec.go]: Yup, I think it should be done in `ContainerExecCreate` where you create `execConfig`.

Comment (calavera) [inline daemon/exec.go]: moved to `ContainerExecCreate`

File: integration-cli/docker_cli_exec_test.go [hunk 2]
```diff
@@ -635,3 +634,35 @@ func (s *DockerSuite) TestExecWithUser(c *check.C) {
...
+		`FROM busybox
+		RUN echo 'dockerio:x:1001:1001::/bin:/bin/false' >> /etc/passwd
+		USER dockerio
+		RUN [ $(whoami) = 'dockerio' ]`,
+		true)
+	if err != nil {
+		c.Fatalf("Could not build image %s: %v", name, err)
+	}
+
+	// run container on first image
```
Comment (LK4D4) [inline integration-cli/docker_cli_exec_test.go]: what this check does?

File: integration-cli/docker_cli_exec_test.go [hunk 3]
```diff
@@ -635,3 +634,35 @@ func (s *DockerSuite) TestExecWithUser(c *check.C) {
...
+	if out, _, err := runCommandWithOutput(exec.Command(dockerBinary, "run", name)); err != nil {
+		c.Fatalf("Could not run image %s: %v: %s", name, err, out)
+	}
+
+	runCmd := exec.Command(dockerBinary, "run", "-d", "--name", "parent", name, "top")
```
Comment (LK4D4) [inline integration-cli/docker_cli_exec_test.go]: I think you can replace all this with `dockerCmd`.

File: integration-cli/docker_cli_exec_test.go [hunk 4]
```diff
@@ -635,3 +634,35 @@ func (s *DockerSuite) TestExecWithUser(c *check.C) {
...
+	if out, _, err := runCommandWithOutput(runCmd); err != nil {
+		c.Fatal(out, err)
+	}
+
+	cmd := exec.Command(dockerBinary, "exec", "parent", "id")
```
Comment (LK4D4) [inline integration-cli/docker_cli_exec_test.go]: `whoami` probably will be simpler

**Comment by calavera** [comment]
@LK4D4 I've cleaned the test. There was a bunch of copy/paste from the builder tests that was not necessary.

File: integration-cli/docker_cli_exec_test.go [hunk 5]
```diff
@@ -635,3 +633,26 @@ func (s *DockerSuite) TestExecWithUser(c *check.C) {
...
+		RUN echo 'dockerio:x:1001:1001::/bin:/bin/false' >> /etc/passwd
+		USER dockerio`,
+		true)
+	if err != nil {
+		c.Fatalf("Could not build image %s: %v", name, err)
+	}
+
+	dockerCmd(c, "run", "-d", "--name", "dockerioexec", name, "top")
+
+	cmd := exec.Command(dockerBinary, "exec", "dockerioexec", "whoami")
```
Comment (LK4D4) [inline integration-cli/docker_cli_exec_test.go]: actually `dockerCmd` returns output, so you can replace this too.

Comment (calavera) [inline integration-cli/docker_cli_exec_test.go]: yeah, I don't know why the other tests preserve this call. Changing anyways.

**Comment by cpuguy83** [comment]
LGTM

**Comment by LK4D4** [comment]
LGTM

**Comment by calavera** [comment]
Cherry picked into the 1.7.1 release branch.

**Comment by michaelsauter** [comment]
Thanks guys :+1:

**Pull request merged**